### PR TITLE
Add Workers Ai types

### PR DESCRIPTION
**Pull request by G4brym**

### CODE REVIEW

**Review by dom96 — APPROVED** [review]

File: types/defines/ai.d.ts [hunk 1]
```diff
@@ -0,0 +1,1022 @@
...
+declare type TensorOpts = {
+  shape?: number[];
+  name?: string;
+  validate?: boolean;
+};
+declare type TensorsObject = {
+  [name: string]: Tensor<any>;
+};
+declare const TypedArrayProto: any;
+declare function isArray(value: any): boolean;
```
Comment (penalosa) [inline types/defines/ai.d.ts:922]: Irrelevant methods like this shouldn't be included

**Review by penalosa — CHANGES_REQUESTED** [review]

File: types/defines/ai.d.ts (hunk 1 above)
Comment (G4brym) [reply types/defines/ai.d.ts]: Unused/irrelevant methods removed 👍

File: types/defines/ai.d.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    ? ModelMappings[K]["class"]
+    : never;
+}[keyof ModelMappings];
+declare type GetModelInstanceType<M extends ModelName> = InstanceType<
+  GetModelClassType<M>
+>;
+declare type GetPostProcessedOutputsType<M extends ModelName> =
+  GetModelInstanceType<M>["postProcessedOutputs"];
+declare type GetInputsType<M extends ModelName> =
+  GetModelInstanceType<M>["inputs"];
```
[the remark this answers is not in the document]
Comment (G4brym) [reply types/defines/ai.d.ts:278]: yes, this helpers are to enable auto complete in the modelName and to auto complete the model inputs based on the modelName the user picked, it also does auto complete for the return type of the .run method

File: types/defines/ai.d.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+declare type AiOptions = {
+  debug?: boolean;
+  prefix?: string;
+  extraHeaders?: object;
+};
+declare class InferenceUpstreamError extends Error {
+  constructor(message: string);
+}
+declare class Ai {
```
[the remark this answers is not in the document]
Comment (G4brym) [reply types/defines/ai.d.ts]: Updated the class to be abstract

File: types/defines/ai.d.ts [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+type AiOptions = {
+  prefix?: string;
+  extraHeaders?: object;
+};
+declare class InferenceUpstreamError extends Error {
+  constructor(message: string);
+}
+declare abstract class Ai {
+  private readonly fetcher;
+  private options;
```
Comment (penalosa) [inline types/defines/ai.d.ts:280]: This are internal implementation details that shouldn't be included

File: types/defines/ai.d.ts [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  lastRequestId: string | null;
+  constructor(fetcher: Fetcher);
```
Comment (penalosa) [inline types/defines/ai.d.ts:282]: This isn't user constructable, so the constructor shouldn't be included

**Review by penalosa — COMMENTED** [review]
Couple comments, and it still looks like there are some none abstract classes—are those intended to be user constructable?

**Review by penalosa — APPROVED** [review]

**Pull request merged**